tasks/build.py

```diff
@@ -346,21 +346,60 @@ def download_pr(repo_name, branch_name, pr, arch_job_dir):
+    if clone_exit_code != 0:
+        return clone_output, clone_error, clone_exit_code
 
     git_checkout_cmd = ' '.join([
         'git checkout',
         branch_name,
     ])
+    print(f'checking out with command {git_checkout_cmd}')
     checkout_output, checkout_err, checkout_exit_code = run_cmd(git_checkout_cmd,
-                                                                "checkout branch '%s'" % branch_name, arch_job_dir)
+                                                                "checkout branch '%s'" % branch_name, arch_job_dir, raise_on_error=False)
+    if checkout_exit_code != 0:
+        return  checkout_output, checkout_err, checkout_exit_code
 
     curl_cmd = f'curl -L https://github.com/{repo_name}/pull/{pr.number}.diff > {pr.number}.diff'
-    curl_output, curl_error, curl_exit_code = run_cmd(curl_cmd, "Obtain patch", arch_job_dir)
-
+    print(f'curl with command {curl_cmd}')
+    curl_output, curl_error, curl_exit_code = run_cmd(curl_cmd, "Obtain patch", arch_job_dir, raise_on_error=False)
+    if curl_exit_code != 0:
+        return  curl_output, curl_error, curl_exit_code
+    
     git_apply_cmd = f'git apply {pr.number}.diff'
-    git_apply_output, git_apply_error, git_apply_exit_code = run_cmd(git_apply_cmd, "Apply patch", arch_job_dir)
+    print(f'git apply with command {git_apply_cmd}')
+    git_apply_output, git_apply_error, git_apply_exit_code = run_cmd(git_apply_cmd, "Apply patch", arch_job_dir, raise_on_error=False)
+    if git_apply_exit_code != 0:
+        return git_apply_output, git_apply_error, git_apply_exit_code
+    
+
+def comment_download_pr(base_repo_name, pr, download_pr_exit_code, download_pr_error):
+    """
+    Handle download_pr() exit code and write helpful comment to PR in case of failure
+
+    Args:
+        base_repo_name (string): name of the repository (format USER_OR_ORGANISATION/REPOSITORY)
+        pr (github.PullRequest.PullRequest): instance representing the pull request
+        download_pr_exit_code (int): exit code from download_pr(). 0 if all tasks were successful, 
+            otherwise it corresponds to the error codes of git clone, git checkout, git apply, or curl.  
+        download_pr_error (string): none, or the output of stderr from git clone, git checkout, git apply or curl. 
 
+    Return:
+        None (implicitly). A comment is created in the appropriate PR.
+        
+    """
+    if download_pr_exit_code != 0:
+        fn = sys._getframe().f_code.co_name
+        download_comment = (f"`{download_pr_error}`"
+                           f"\nUnable to download or merge changes between the source branch and the destination branch.\n"
+                           f"\nTip: This can usually be resolved by syncing your branch and resolving any merge conflicts.")
+        download_comment = pr_comments.create_comment(repo_name=base_repo_name, pr_number=pr.number, comment=download_comment)
```

Comment:
This creates a new comment (for each job that failed), right? So we might get many new comments. Wonder if it could be possible to pass the result back to the calling function such that it is included in the responses to the `bot: build` command(s)?

Reply:
That's a good point. I actually meant for it to just comment once and stop with the exception, but I did not actually verify that it works that way because when I was playing around with it I was only building for one target: https://github.com/Neves-Bot/software-layer/pull/12#issuecomment-1814142716

In that case of course the bot could only write one comment for that one target, so I'm completely sure the raising the error is enough 🤔